src/test/java/com/xwiki/authentication/saml/XWikiGroupManagerTest.java

```diff
@@ -0,0 +1,164 @@
+package com.xwiki.authentication.saml;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.classes.BaseClass;
+import com.xpn.xwiki.web.XWikiRequest;
+import com.xpn.xwiki.web.XWikiResponse;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.slf4j.LoggerFactory;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+
+import javax.servlet.http.HttpSession;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class XWikiGroupManagerTest {
+    @Test
+    public void addUserToGroup_ShouldAddUserToGivenGroup() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+        .whenAddedToGroup()
+        .then()
+            .user("ArthurDent")
+            .isInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void removeUserFromGroup_ShouldRemoveUserFromGivenGroup() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+            .userIsAlreadyInGroup("ArthurDent","StarshipTroopers")
+        .whenRemoveFromGroup()
+        .then()
+            .user("ArthurDent")
+            .isntInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void userAlreadyMemberOfGroup_ShouldIgnoreTheNewMembership() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+            .userIsAlreadyInGroup("ArthurDent","StarshipTroopers")
+        .whenAddedToGroup()
+        .then()
+            .user("ArthurDent")
+            .isInGroup("StarshipTroopers");
+    }
+
+    @Test
+    public void removeUserFromGroupWhereHeIsnt_ShouldNotChangeGroupMembership() {
+        given()
+            .user("ArthurDent")
+            .fromGroup("StarshipTroopers")
+        .whenRemoveFromGroup()
+        .then()
+            .user("ArthurDent")
+            .isntInGroup("StarshipTroopers");
+    }
+
```

Comment:
Am missing two tests related to adding or removing to a group not affecting the remaining groups. I'd write then like this:
1. Adding test:
    * Given a user in a group.
    * When the user is added to another group.
    * Then the user is part of both groups.
 2. Removing Test:
    * Given a user in two groups.
    * When the user is removed from the first group.
    * Then the user is still a part of the second group.